dbt/parser/base.py

```diff
@@ -117,10 +118,23 @@ def parse_node(cls, node, node_path, root_project_config,
         runtime_config = db_wrapper.config
         adapter.release_connection(parsed_node.name)
 
-        # Special macro defined in the global project
+        # Special macro defined in the global project. Use the root project's
+        # definition, not the current package
         schema_override = config.config.get('schema')
-        get_schema = context.get('generate_schema_name',
-                                 lambda x: default_schema)
+        get_schema_macro = macro_manifest.find_macro_by_name(
+            'generate_schema_name',
+            root_project_config.project_name
+        )
+        if get_schema_macro is None:
+            get_schema_macro = macro_manifest.find_macro_by_name(
+                'generate_schema_name',
+                dbt.include.GLOBAL_PROJECT_NAME
+            )
+        if get_schema_macro is None:
+            def get_schema(_):
+                return default_schema
+        else:
+            get_schema = get_schema_macro.generator(context)
```

Comment:
I think there's an issue with using the node's `context` here. Given a custom schema override that looks like this:

```jinja
{% macro some_local_macro(custom_schema_name=none) -%}
  abc
{%- endmacro %}

{% macro generate_schema_name(custom_schema_name=none) -%}
  {{ some_local_macro() }}
{%- endmacro %}
```

The `generate_schema_name` macro is plucked out of the `macro_manifest` explicitly, but the `some_local_macro` macro is local to the root project. The `generate_schema_name` macro is ripped away from the rest of its context, so models in an imported package will fail with `some_local_macro` is not defined.

One way to circumvent this is to namespace the macro. This works:
```
{% macro generate_schema_name(custom_schema_name=none) -%}
  {{ my_project_name.some_local_macro() }}
{%- endmacro %}
```

The `my_project_name` dict is defined in the dependency's context, and it does indeed contain a function called `some_local_macro`.

I'm unsure of the best way to address this. I think we have a couple of options (or some combo thereof):
1. require macro calls in `generate_schema_name` to be namespaced (unsure if we have a mechanism to do this)
2. throw an informative error if the `generate_schema_name` macro fails to compile, possibly linking to docs
3. use the "root" context when building the `get_schema_macro` generator. I did this as an experiment, and it seemed to work ok, but I don't love it.

Approach 3 looked something like:
```python
        parsed_node = ParsedNode(**node)

        fake_node = node.copy()
        fake_node['package_name'] = root_project_config.project_name
        fake_parsed_node = ParsedNode(**fake_node)

        context = dbt.context.parser.generate(parsed_node, root_project_config,
                                              macro_manifest, config)

        root_context = dbt.context.parser.generate(fake_parsed_node, root_project_config,
                                              macro_manifest, root_project_config)

```

I guess `dbt.context.parser.generate` cares about the `package_name` field of the node it's generating a context for. Didn't dig too much further into it though.

Reply:
@beckjake yeah, I'm 100% with you. Would be good to avoid the cost of doing the same thing for every model, but also seems like it would be a lot simpler to consider this a full-on special case.